src/lib/openssl_preload/crypto.c

```diff
@@ -0,0 +1,232 @@
+    // Store at the first empty cache slot.
+    for (int i = 0; i < EVP_BACKTRACE_CACHE_LEN; i++) {
+        if (evp_backtrace_cache[i].addr == NULL) {
+            evp_backtrace_cache[i].addr = addr;
+            evp_backtrace_cache[i].is_libssl = is_libssl;
+
+            debuglog("Cached EVP_EncryptUpdate caller=%p, is_libssl=%s\n", addr,
+                     is_libssl ? "true" : "false");
+            break;
+        }
+    }
+}
+
+static bool _is_addr_in_libssl(void* addr) {
+    bool found = false;
+
+    // Gets a string containing the full file path to the caller,
+    // e.g. /lib/x86_64-linux-gnu/libssl.so.1.1
+    char** bt_str_buf = backtrace_symbols(&addr, 1);
+
+    if (bt_str_buf != NULL && bt_str_buf[0] != NULL) {
+        found = (strstr(bt_str_buf[0], "libssl.so") != NULL);
+    }
+
+    free(bt_str_buf);
+    return found;
+}
+
+int EVP_EncryptUpdate(void* cipher, unsigned char* out, int* outl, const unsigned char* in,
+                      int inl) {
+    // In the case of tor:
+    //   - calls from libssl are used for TLS and skipping will break TLS
+    //   - calls from tor are used for AES and can be skipped
+    // So we can skip the crypto op as long as the call is not made from libssl.
+    void* caller_addr = NULL;
+    bool caller_is_libssl = false;
+
+    {
+        // Get the backtrace addresses; we only need 2: our stack is in [0]
+        // and the caller's stack is in [1].
+        void* bt_addr_buf[2];
+
+        // We want to make sure we get the backtrace here and not in a helper
+        // function so that our stack offset math is correct.
+        int bt_len = backtrace(bt_addr_buf, 2);
+        if (bt_len == 2) {
+            caller_addr = bt_addr_buf[1];
+        }
+    }
+
+    if (caller_addr != NULL) {
+        int result = pthread_mutex_lock(&global_state_lock);
+        assert(result == 0);
+
+        // We use a cache first because checking the name of the library is expensive.
+        const bt_cache_entry_t* entry = _get_cache_entry(caller_addr);
+
+        if (entry != NULL) {
+            // The answer we want is cached.
+            caller_is_libssl = entry->is_libssl;
+        } else {
+            // Fall back to check the caller backtrace symbols.
+            // This might be more expensive than just performing the crypto op, and we might have to
+            // perform the crypto op anyway depending on the result, but we do it anyway to maintain
+            // consistency in behavior.
+            caller_is_libssl = _is_addr_in_libssl(caller_addr);
+            // Cache it if there is space.
+            if (_cache_has_space()) {
+                _append_to_cache(caller_addr, caller_is_libssl);
+            }
+        }
+
+        result = pthread_mutex_unlock(&global_state_lock);
+        assert(result == 0);
+    }
+
+    if (caller_addr != NULL && !caller_is_libssl) {
+        // Skip the crypto in calls made from the application, e.g. tor.
+        _increment(&evp_eu_cnt);
```

Comment:
Optional - right now the lock will be taken and released twice in each call to `EVP_EncryptUpdate` - once to access the cache and once to access the counter. We could require that the caller  has already taken the lock before calling _increment, and then here continue holding the lock until after its called.

Reply:
Done.